### ISSUE TITLE
How to identify the number of mics turned on and which one is speaking

### ISSUE THREAD

**Issue by ullfindsmit** [naudio/NAudio#664]
I am working on a winform app that takes audio from a USB mic and saves it to a file.
This part is working fine currently.

What I am being asked now is to identify how many MICs(Inputs) are enabled on an external device like the [zoom F4](https://www.zoom-na.com/products/field-video-recording/field-recording/zoom-f4-multitrack-field-recorder)  and when someone is talking in to one of them (and which one)

In other words, if this device was hooked up to a laptop and there were 4 people sitting around it, assign each mic to the person and know when he or she spoke.

Thanks in Advance.

**Comment by robertkjr3d** [comment]
I pointed out this problem like, oh must have been 5 to 8 years ago.  I thought they would have fixed it by now.  Guess not.  I updated the code to the latest, still not fixed.  Useless!  Must find another code dll.

**Comment by robertkjr3d** [comment]
I pointed out this problem of the device only showing 2-CH like 5 to 8 years ago.  But is still not fixed in the latest version.
We must search for another DLL.

**Comment by markheath** [comment]
Closing this old issue as it's basically about a limitation of the old Windows WinMM ("WaveIn") API that `WaveInCapabilities.Channels` is built on. That value comes straight from the Win32 `WAVEINCAPS.wChannels` field, which was never designed to report the true channel count of multi-input interfaces like your Zoom F4 - you'll see the note in the NAudio source itself: *"Seems to be set to -1 on a lot of devices"* (and in your case it's reporting 2). NAudio is just surfacing what WinMM hands it.

To get the real channel count on Windows, use WASAPI / Core Audio instead. Enumerate with `MMDeviceEnumerator`, then read the endpoint's mix format:

```csharp
using var enumerator = new MMDeviceEnumerator();
foreach (var device in enumerator.EnumerateAudioEndPoints(DataFlow.Capture, DeviceState.Active))
{
    using var client = device.CreateAudioClient();   // NB: CreateAudioClient(), not the obsolete .AudioClient property
    Console.WriteLine($"{device.FriendlyName}: {client.MixFormat.Channels} channels");
}
```

That mix format reflects the endpoint's actual channel layout, which is why CSCore (which reads the same Core Audio API) reports 6.

For your real goal - capturing all 6 inputs separately and detecting who's talking - you've got two paths:

- **WASAPI capture** (`WasapiCapture`, or `WasapiRecorder` in NAudio 3): if the device's shared mix format is 6-channel, this captures all six channels interleaved. De-interleave the buffer and meter each channel to tell who's speaking. Exclusive mode works too if the driver advertises the full 6-channel format.
- **ASIO** (`AsioOut`, or the new `AsioDevice` with `InitRecording`/`InitDuplex` in NAudio 3): the most reliable route for discrete per-input capture at low latency, assuming the F4 ships an ASIO driver.
